### PR TITLE
chore: Decouple backup/restore from `WireCryptobox` - WPB-7395

### PR DESCRIPTION
**Pull request by samwyndham**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-7395" title="WPB-7395" target="_blank">StoryWPB-7395</a>  Move backup/restore encryption out of Cryptobox
  </td></table>
  <br />
 

### Issue

This PR decouples backup & restore from `WireCryptobox`. This is done because we will soon be removing `WireCryptobox` from our project. Specifically it:

1. Introduces to new targets in the `WireFoundation` package:
    - `Clibsodium` which is a binary target that fetches an `XCFramework` from our `Clibsodium` fork.
    - `WireCrypto` which will be used to store various crypto utilities.
2. Moves crypto backup/restore code from `WireCryptobox` to `WireCrypto`.
3. Moves various crypto utilities from `WireFoundation` to `WireCrypto`.
4. Removes `Clibsodium` headers from `WireCryptobox` as these are no longer needed 🥳.
5. Updates `Carthage` not to download `Clibsodium`.

### Testing

Run automated tests

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

### CODE REVIEW

File: WireFoundation/Package.swift [hunk 1]
```diff
@@ -7,6 +7,8 @@ let package = Package(
     name: "WireFoundation",
     platforms: [.iOS(.v16), .macOS(.v12)],
     products: [
+        .library(name: "Clibsodium", targets: ["Clibsodium"]),
```
Comment (samwyndham) [inline WireFoundation/Package.swift:11]: We can remove this as a product once we remove `WireCryptobox`

**Review by johnxnguyen — APPROVED** [review]
Amazing!

File: WireFoundation/Package.swift (hunk 1 above)
Comment (netbe) [reply WireFoundation/Package.swift:11]: might be good to left this comment here as a 
```suggestion
        // TODO: [WPB-7394] `Clibsodium` is no longer needed as a product
        .library(name: "Clibsodium", targets: ["Clibsodium"]),
```

(https://wearezeta.atlassian.net/browse/WPB-7394)]

File: Cartfile [hunk 2]
```diff
@@ -1,5 +1,4 @@
 github "wireapp/cryptobox-ios" "v1.1.0_xcframework_arm64simulator"
-github "wireapp/libsodium" "1.0.14"
```
Comment (netbe) [inline Cartfile:2]: one less here 🎊

File: WireAPI/Sources/WireAPI/Authentication/CookieStorage.swift [hunk 3]
```diff
@@ -17,6 +17,7 @@
 //
 
 import Foundation
+import WireCrypto
```
Comment (netbe) [inline WireAPI/Sources/WireAPI/Authentication/CookieStorage.swift:20]: why WireAPI needs WireCrypto?

File: WireFoundation/Tests/TestPlans/AllTests.xctestplan [hunk 4]
```diff
@@ -17,8 +17,8 @@
     {
       "target" : {
         "containerPath" : "container:WireFoundation",
-        "identifier" : "WireLoggingTests",
-        "name" : "WireLoggingTests"
+        "identifier" : "WireCryptoTests",
```
Comment (netbe) [inline WireFoundation/Tests/TestPlans/AllTests.xctestplan:20]: 👍

**Review by netbe — APPROVED** [review]
Looks good, @samwyndham let's wait for 3.115.0 to be branched out (today) before merging this

File: WireAPI/Sources/WireAPI/Authentication/CookieStorage.swift (hunk 3 above)
Comment (samwyndham) [reply WireAPI/Sources/WireAPI/Authentication/CookieStorage.swift:20]: It uses `static AES256Crypto.encryptAllAtOnceWithPrefixedIV(plaintext:key:)` which was in `WireFoundation` but moves to `WireCrypto` in this PR.

Comment (samwyndham) [reply WireAPI/Sources/WireAPI/Authentication/CookieStorage.swift:20]: @netbe I'm resolving this comment. Feel free to re-open if you would like to discuss further.

**Pull request merged**